File: lib/contrek/finder/concurrent/position.rb
module Contrek
  module Concurrent
    class Position
      include Listable

      attr_reader :end_point

      def initialize(hub:, position:, known_endpoint: nil)
        if !known_endpoint.nil?
          @end_point = known_endpoint
          @position = @end_point.position
        else
          @end_point = hub.payloads[position[:y]] ||= EndPoint.new(position) if hub
          @position = position
        end
      end

      def payload
        @position
      end

      def after_add(new_queue)
        if @end_point && new_queue.instance_of?(Contrek::Concurrent::Part)
          @end_point.queues << new_queue if !@end_point.queues.include?(new_queue)
          if @end_point.queues.size > 1
            new_queue.polyline.any_ancients = true
            @end_point.queues.first.polyline.any_ancients = true
          end
        end
      end

      def before_rem(old_queue)
        # @end_point&.queues&.delete(old_queue) if old_queue.class == Contrek::Concurrent::Part
      end

      def inspect
        "#{self.class} (#{payload})"
      end
    end
  end
end

File: ext/cpp_polygon_finder/PolygonFinder/src/polygon/finder/concurrent/Position.cpp
/*
 * Position.cpp
 *
 * Copyright (c) 2025-2026 Emanuele Cesaroni
 *
 * Licensed under the GNU Affero General Public License v3 (AGPLv3).
 * See the LICENSE file in this directory for the full license text.
 */
#include <iostream>
#include <algorithm>
#include "Position.h"
#include "Hub.h"
#include "Part.h"
#include "Polyline.h"

Position::Position(Hub* hub, Point* point)
: QNode<Point>(point)
{ if (hub != nullptr) {
    int key = point->y;
    EndPoint* existing_ep = hub->get(key);
    if (existing_ep == nullptr)
    { end_point_ = hub->put(key, hub->spawn_end_point());
      end_point_->set_point(point);
    } else {
      end_point_ = existing_ep;
    }
  }
}

Position::Position(EndPoint* end_point)
  : QNode<Point>(end_point->get_point())
{  this->end_point_ = end_point;
}

void Position::before_rem(Queueable<Point>* q)  {
  /*if (this->end_point_ != nullptr)
  { if (q->listable())
    { auto& v = this->end_point_->queues();
      v.erase(std::remove(v.begin(), v.end(), q), v.end());
    }
  }*/
}

void Position::after_add(Queueable<Point>* q) {
  if (this->end_point_ != nullptr && q->listable()) {
    auto& q_vec = this->end_point_->queues();
    auto it = std::find(q_vec.begin(), q_vec.end(), q);
    if (it == q_vec.end()) {
      q_vec.push_back(q);
      if (q_vec.size() > 1) {
        static_cast<Part*>(q)->polyline()->any_ancients = true;
        Queueable<Point>* first_q = q_vec.front();
        static_cast<Part*>(first_q)->polyline()->any_ancients = true;
      }
    }
  }
}
